### PR TITLE
Add page on relation between Gardener and Cluster API

### PR DESCRIPTION
**Pull request by vlerenc**
**How to categorize this PR?**
/area documentation

**What this PR does / why we need it**:
Add page on relation between Gardener and Cluster API to help people that ask questions how both are related to each other or why Gardener has its own API.

**Special notes for your reviewer**:
/invite @rfranzke @vasu1124 @hardikdr

### CODE REVIEW

**Comment by vlerenc** [comment]
> Should we make this be more prominent? It's an obvious question of somebody that knows/has heard of cluster API, i.e., "Gardener sounds very similar to cluster API - what's actually the difference"?

I don't know @rfranzke . At least we have now a statement. What do you suggest?

> Can you also add the document to the index (https://github.com/gardener/gardener/blob/master/docs/README.md)?

Sure, done.

> s/But because/But, or?

Well, I wanted to stress it a bit more. Not OK?

As for the language itself, I think it's common, see e.g. in code similar comments (whether by native speakers or not I haven't checked):
```
vendor/golang.org/x/net/http2/transport.go
762:// only be called when we're idle, but because we're coming from a new

vendor/golang.org/x/net/http2/server.go
2389:// writeChunk writes chunks from the bufio.Writer. But because
```

**Comment by rfranzke** [comment]
> I don't know @rfranzke . At least we have now a statement. What do you suggest?

I was thinking of a dedicated section in the README.md with a header like "Comparison to cluster API" or something similar (now it's just a sentence that is "hard to find".

> Well, I wanted to stress it a bit more. Not OK?

OK, no, if it's on purpose then everything is good :-) Just wanted to clarify on that. Thanks!

File: docs/concepts/cluster-api.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+To wrap it up, while the Cluster API knows about clusters, it doesn't know about their make-up. With Gardener, we wanted to go beyond that and harmoise the make-up of the clusters themselves and make them homogenous across all supported infrastructures. Gardener can therefore deliver homogenous clusters with exactly the same configuration and behaviour on all infrastructures (see also [Gardener's coverage in the official conformance test grid](https://k8s-testgrid.appspot.com/conformance-all)).
```
Comment (hardikdr) [inline docs/concepts/cluster-api.md]:
```suggestion
To wrap it up, while the Cluster API knows about clusters, it doesn't know about their make-up. With Gardener, we wanted to go beyond that and harmonise the make-up of the clusters themselves and make them homogenous across all supported infrastructures. Gardener can therefore deliver homogenous clusters with exactly the same configuration and behaviour on all infrastructures (see also [Gardener's coverage in the official conformance test grid](https://k8s-testgrid.appspot.com/conformance-all)).
```

File: docs/concepts/cluster-api.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+With [Cluster API `v1alpha3`](https://kubernetes.io/blog/2020/04/21/cluster-api-v1alpha3-delivers-new-features-and-an-improved-user-experience) and the support for declarative control plane management, it became now possible (again) to enable Kubernetes managed services like GKE or Gardener. We would be more than happy, if the community would be interested to contribute a Gardener control plane provider.
```
Comment (wyb1) [inline docs/concepts/cluster-api.md]:
```suggestion
With [Cluster API `v1alpha3`](https://kubernetes.io/blog/2020/04/21/cluster-api-v1alpha3-delivers-new-features-and-an-improved-user-experience) and the support for declarative control plane management, it became now possible (again) to enable Kubernetes managed services like GKE or Gardener. We would be more than happy, if the community would be interested, to contribute a Gardener control plane provider.
```

File: docs/concepts/cluster-api.md (hunk 1 above)
Comment (wyb1) [inline docs/concepts/cluster-api.md]: First paragraph uses `harmonize` while this paragraph uses `harmonise`. Choose either American or British english 😄

File: docs/concepts/cluster-api.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Relation between Gardener API and Cluster API (SIG Cluster Lifecycle)
+
+In essence, the Cluster API harmonizes the way how to get to clusters, while Gardener goes one step further and harmonizes also the clusters themselves. The Cluster API delegates the specifics to so-called providers for infrastructures or control planes via specific CR(D)s while Gardener only has one cluster CR(D). Different Cluster API providers, e.g. for AWS, Azure, GCP, etc. give you vastly different Kubernetes clusters. In contrast, Gardener gives you the exact same clusters with the exact same K8s version, operating system, control plane configuration like for API server or kubelet, add-ons like overlay network, HPA/VPA, DNS and certificate controllers, ingress and network policy controllers, control plane monitoring and logging stacks, down to the behavior of update procedures, auto-scaling, self-healing, etc. on all supported infrastructures. This, homogenous clusters, is an essential goal for Gardener as its main purpose is to simplify operations for teams that need to develop and ship software on Kubernetes clusters on a plethora of infrastructures (a.k.a. multi-cloud).
```
Comment (wyb1) [inline docs/concepts/cluster-api.md]:
```suggestion
In essence, the Cluster API harmonizes how to get to clusters, while Gardener goes one step further and also harmonizes the clusters themselves. The Cluster API delegates the specifics to so-called providers for infrastructures or control planes via specific CR(D)s while Gardener only has one cluster CR(D). Different Cluster API providers, e.g. for AWS, Azure, GCP, etc. give you vastly different Kubernetes clusters. In contrast, Gardener gives you the exact same clusters with the exact same K8s version, operating system, control plane configuration like for API server or kubelet, add-ons like overlay network, HPA/VPA, DNS and certificate controllers, ingress and network policy controllers, control plane monitoring and logging stacks, down to the behavior of update procedures, auto-scaling, self-healing, etc. on all supported infrastructures. These homogenous clusters are an essential goal for Gardener as its main purpose is to simplify operations for teams that need to develop and ship software on Kubernetes clusters on a plethora of infrastructures (a.k.a. multi-cloud).
```

**Comment by vlerenc** [comment]
Thanks @rfranzke @hardikdr @wyb1, anything else?

As for your point @rfranzke, the `README` looked like you didn't want to go too much into details and promote one topic over the other and keep it concise. Anyway, I can propose that, maybe as a separate PR if you really want that topic features in the `README`.

**Review by rfranzke — DISMISSED** [review]
I'm okay with it for now, thanks!

/lgtm

**Review by hardikdr — DISMISSED** [review]
/lgtm

**Review by wyb1 — DISMISSED** [review]

**Comment by vlerenc** [comment]
@rfranzke Would you then please merge, then I would start using it in some replies. Thanks in advance!

**Comment by rfranzke** [comment]
Sure, does @vasu1124 plan to review it as well? If not/not soon then let's get it in 👀

**Comment by rfranzke** [comment]
@vasu1124 mentioned that he plans to review, so let's wait for him a bit

File: docs/concepts/cluster-api.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# Relation between Gardener API and Cluster API (SIG Cluster Lifecycle)
+
+In essence, the Cluster API harmonizes how to get to clusters, while Gardener goes one step further and also harmonizes the clusters themselves. The Cluster API delegates the specifics to so-called providers for infrastructures or control planes via specific CR(D)s while Gardener only has one cluster CR(D). Different Cluster API providers, e.g. for AWS, Azure, GCP, etc. give you vastly different Kubernetes clusters. In contrast, Gardener gives you the exact same clusters with the exact same K8s version, operating system, control plane configuration like for API server or kubelet, add-ons like overlay network, HPA/VPA, DNS and certificate controllers, ingress and network policy controllers, control plane monitoring and logging stacks, down to the behavior of update procedures, auto-scaling, self-healing, etc. on all supported infrastructures. These homogenous clusters are an essential goal for Gardener as its main purpose is to simplify operations for teams that need to develop and ship software on Kubernetes clusters on a plethora of infrastructures (a.k.a. multi-cloud).
```
Comment (vasu1124) [inline docs/concepts/cluster-api.md]: s/homogenous/homogeneous

File: docs/concepts/cluster-api.md [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+That means, we follow the [Cluster API](https://github.com/kubernetes-sigs/cluster-api#cluster-api) with great interest and are active members. It was completely overhauled from `v1alpha1` to `v1alpha2`. But because `v1alpha2` made too many assumptions about the bring-up of masters and was enforcing master machine operations (see [here](https://cluster-api.sigs.k8s.io/user/concepts.html#control-plane): “As of `v1alpha2`, Machine-Based is the only control plane type that Cluster API supports”), services that managed their control planes differently like GKE or Gardener couldn't adopt it (e.g. [Google only supports `v1alpha1`](https://cloud.google.com/anthos/gke/docs/on-prem/concepts/cluster-api)). In 2020 [`v1alpha3`](https://kubernetes.io/blog/2020/04/21/cluster-api-v1alpha3-delivers-new-features-and-an-improved-user-experience/) was introduced and made it possible (again) to integrate managed services like GKE or Gardener. The mapping from the Gardener API to the Cluster API is mostly syntactic.
+
```
Comment (vasu1124) [inline docs/concepts/cluster-api.md:8]:
```suggestion

On the implementation history: the Gardener development did not start tackling the bootstrap problem first. It started with the inception design and made the bold assumption that there is already one Kubernetes work plane to bootstrap from. This accelerated the development of Gardener on the basis of an already existing Kubernetes substrate and skills, i.e. launching, scaling managing all components as pods, which made Gardener quickly ready for production (just like many other applications that kickstarted on Kubernetes). Another way to contextualize the development paths is that while Cluster API needed to tackle the challenge bottom-up, Gardener went top-down. Ultimately, both approaches end up having similar requirements in the lowest execution layer and indeed have crosspollinated each other. And in Kubernetes terms, both approches are in a _reconciliation loop_.
```

**Review by vasu1124 — CHANGES_REQUESTED** [review]

**Comment by majst01** [comment]
Hi @vlerenc,

nice write up, i would also vote to bring this more prominent to the README.md as @rfranzke already said.
Having it in the last paragraph in the last word is not enough. I would rather create a new paragraph in the README.md which will summarize and tease the content.

File: docs/concepts/cluster-api.md (hunk 4 above)
Comment (vlerenc) [reply docs/concepts/cluster-api.md]: Fixed by 2c7e849dfaef289c0b670a780506f4e3ead92d44.

File: docs/concepts/cluster-api.md (hunk 5 above)
Comment (vlerenc) [reply docs/concepts/cluster-api.md:8]: Hmm... I don't like prolonging this paragraph with the Cluster APIs bootstrap problem or Gardener's implementation history and introducing a bunch of new terms like said bootstrapping, inception, pod-based control planes (you will have noticed that I deliberately didn't go into this area), substrate, bottom-up or top-down that even confuses me here including the cross-pollination ;-), and finally what you mean behind reconciliation.

In other words, while I understand what you are saying, that's not my text and what I wanted to express here and I rather suggest for you to file another PR instead. I hope, that's OK with you/you won't be mad at me, all good, just a different angle.

**Comment by vlerenc** [comment]
@majst01 OK, thanks, so two votes for a paragraph in the top-level README, so I'll try to write something for that as well.

/hold

File: docs/concepts/cluster-api.md (hunk 5 above)
Comment (vasu1124) [reply docs/concepts/cluster-api.md:8]: /lgtm then. But I do think this angle makes the relationship much clearer.

**Review by vasu1124 — DISMISSED** [review]

**Comment by vlerenc** [comment]
Added paragraph on cluster API to the top-level README with b0dd9ace13851f23558bf3ed4db6bfc7b259860b. Please merge if you agree. From my side:
/lgtm

**Review by rfranzke — DISMISSED** [review]
Thanks @vlerenc 
/lgtm
/unhold

**Comment by vlerenc** [comment]
Where is the UI showing the review state? When I `curl` the issue I see that @vasu1124 is still in `CHANGES_REQUESTED`, which is why @neo-liang-sap [reviews bot]( https://github.tools.sap/kubernetes/gardener-robot/blob/master/robot/bots/reviews.py) continuously puts the PR back to `needs/changes`. That's not really productive, is it?

/uninvite @vasu1124 
/lgtm

**Review by rfranzke — APPROVED** [review]

**Comment by vlerenc** [comment]
I don't understand. All conversations were resolved, only @rfranzke 's review was listed as approved, I saw no requested changes by @vasu1124 **in the UI**. Yet, when calling the API, I see:
```
    "id": 534154395,
    "node_id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NTM0MTU0Mzk1",
    "user": {
      "login": "vasu1124",
      "id": 11454395,
      "node_id": "MDQ6VXNlcjExNDU0Mzk1",
      "avatar_url": "https://avatars3.githubusercontent.com/u/11454395?u=9496275f0718d81bea32068dccb4d9c2d848e592&v=4",
      "gravatar_id": "",
      "url": "https://api.github.com/users/vasu1124",
      "html_url": "https://github.com/vasu1124",
      "followers_url": "https://api.github.com/users/vasu1124/followers",
      "following_url": "https://api.github.com/users/vasu1124/following{/other_user}",
      "gists_url": "https://api.github.com/users/vasu1124/gists{/gist_id}",
      "starred_url": "https://api.github.com/users/vasu1124/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/vasu1124/subscriptions",
      "organizations_url": "https://api.github.com/users/vasu1124/orgs",
      "repos_url": "https://api.github.com/users/vasu1124/repos",
      "events_url": "https://api.github.com/users/vasu1124/events{/privacy}",
      "received_events_url": "https://api.github.com/users/vasu1124/received_events",
      "type": "User",
      "site_admin": false
    },
    "body": "",
    "state": "CHANGES_REQUESTED",
    "html_url": "https://github.com/gardener/gardener/pull/3181#pullrequestreview-534154395",
    "pull_request_url": "https://api.github.com/repos/gardener/gardener/pulls/3181",
    "author_association": "MEMBER",
    "_links": {
      "html": {
        "href": "https://github.com/gardener/gardener/pull/3181#pullrequestreview-534154395"
      },
      "pull_request": {
        "href": "https://api.github.com/repos/gardener/gardener/pulls/3181"
      }
    },
    "submitted_at": "2020-11-19T08:09:22Z",
    "commit_id": "ce543c0d2039d18813fa27303818721d403555df"
```
I.e. the review still is in state `CHANGES_REQUESTED` and that's why the bot continuous to mess up the labels.

Where you aware (cc @neo-liang-sap) or am I confused here/am I missing something?

**Comment by vlerenc** [comment]
Thanks for merging anyway @rfranzke .
